pennylane/tape/qscript.py

```diff
@@ -1410,3 +1411,76 @@ def wrapper(*args, **kwargs):
         return qscript
 
     return wrapper
+
+
+def bind_new_parameters_tape(
+    tape: QuantumScript, params: Sequence[TensorLike], indices: Sequence[int]
+):
+    """Create a new tape with updated parameters.
+
+    This function takes a :class:`~.tape.QuantumScript` as input, and returns
+    a new ``QuantumScript`` containing the new parameters at the provided indices,
+    with the parameters at all other indices remaining the same.
+
+    Args:
+        tape (.tape.QuantumScript): Tape to update
+        params (Sequence[TensorLike]): New parameters to create the tape with. This
+            must have the same length as ``indices``.
+        indices (Sequence[int]): The parameter indices to update with the given parameters.
+            The index of a parameter is defined as its index in ``tape.get_parameters()``.
+
+    Returns:
+        .tape.QuantumScript: New tape with updated parameters
+    """
+    # pylint: disable=no-member
+
+    if len(params) != len(indices):
+        raise ValueError("Number of provided parameters does not match number of indices")
+
+    new_ops = []
+    idx = 0
+    p_idx = 0
+
+    for op in tape.circuit:
```

Comment:
I just realized that it might be possible to speed up this approach. Rather than iterating through the entire circuit, it makes more sense to just iterate through the indices for which the parameters are being updated. Currently we have O(n) time complexity with n being the number of items in the circuit, and the proposed approach would have complexity O(i) with i being the number of parameters to update. See [here](https://github.com/PennyLaneAI/pennylane/blob/330491ece7d8a2cf2db238c8cfd508bfd7378ae7/pennylane/gradients/general_shift_rules.py#L384) in `qml.gradients`.

Reply:
Hmmm, that makes sense. I can't use `get_operation` here though since that only gets trainable parameters, but there's definitely a way to make this work by using `tape._par_info` directly.

Reply:
Updated 👍